RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,179 @@
+|||Total|138|
+
+We can maintain a mapping of metric key (which can be anywhere between 30 to 50 characters) and metric value in the impelementation of BaseStatsReporter class.
+Each of the above metric value is stored in a 8 byte integer type. At any moment we may have 138 × 8 = 1104 bytes of values data held in-memory. Also, we have 138 × ~50 = 6900 ~ 7K bytes for keys. In total we have 7K + 1104 ~ 8K bytes of memory. Thus holding metrics in memory has least impact on Prestissimo worker.
+Given the above estimates, we decide keep the stats in-memory and we can revisit this if the memory footprint grows.
+
+Note: Estimating the size of a histogram with quantiles is not straight forward, it depends on rotation time and error tolerance settings.
+
+### Metric timestamps
+Since we overwrite metrics in our current design, it is likely that Prometheus may miss updates as it pulls metrics at regular intervals. The timestamps are assigned by Prometheus server (this is recommended) and may not reflect the exact timestamp at which the metric was recorded.
+On other hand, we can maintain a list of metric values and timestamps seen so far in Prestissimo and in the response to the Prometheus we can include this <timestamp, value> pairs. Prometheus must be configured to honor these timestamps. But this approach is not recommended.
+
+### Serialization Format (TBD)
+#### Prometheus Data model
+We need to define metric labels which are used at Prometheus client end to filter metrics. For instance, a simple label in our case could be cluster name and worker IP. Since metrics are coming from each worker, we need a way to isolate and monitor them. For cluster name, we are relying on the node.environment config property and for worker IP, we are relying on the HOSTNAME environment variable. Here is a sample of metrics formatted using this data model.
+```
+# HELP presto_cpp_http_client_presto_exchange_source_num_on_body
+# TYPE presto_cpp_http_client_presto_exchange_source_num_on_body counter
+presto_cpp_http_client_presto_exchange_source_num_on_body{cluster="testing",worker="Local"} 5
+# TYPE presto_cpp_memory_cache_hit_bytes gauge
+presto_cpp_memory_cache_hit_bytes{cluster="testing",worker="Local"} 0
+# HELP presto_cpp_mapped_memory_bytes
+# TYPE presto_cpp_mapped_memory_bytes gauge
+presto_cpp_mapped_memory_bytes{cluster="testing",worker="Local"} 806912
+# HELP presto_cpp_os_system_cpu_time_micros
+# TYPE presto_cpp_os_system_cpu_time_micros gauge
+presto_cpp_os_system_cpu_time_micros{cluster="testing",worker="Local"} 2218
+```
+#### Define Custom interface
+By default, we shall implement JSON and Prometheus Data Model serialization of metrics. We shall expose a new Serialization interface that users can implement to customize serialization.
+Pros: In-house.
+Cons: Keeping it in sync with Prometheus data model.
+
+#### Use Prometheus-CPP:
+Pros: Popular and simplifies histogram and summary metric maintenance.
+Cons: External dependency.
+
+
+## Configuration
```

Comment:
Can you complete this section before Thursday's session ? Its important to ensure this is simple.